File: src/framework/web/pages/admin-page.ts
import type { Page } from '@playwright/test';
import { LoggerUtil } from '../../utils/logger.js';
import { formatLocator } from '../../utils/locator.js';
import { BasePage } from '../base-page.js';

export class AdminPage extends BasePage {
  private static readonly DELETE_SELECTED_BTN = "//button[contains(@class, 'label-danger')]";
  private static readonly MODAL_CANCEL_BTN = "//div[contains(@class, 'modal-footer')]/button[contains(@class, 'ghost')]";
  private static readonly USERNAME_CHK = "//div[@class='oxd-table-card'][.//div[text()='%s'] and .//div[text()='%s']]//span[contains(@class,'checkbox-input')]";

  constructor(page: Page) {
    super(page);
  }

  public async selectFirstUser(role: string, status: string): Promise<boolean> {
    try {
      const candidateRow = this.page
        .locator("div.oxd-table-card")
        .filter({ hasText: role })
        .filter({ hasText: status })
        .first();

      const count = await candidateRow.count();
      if (count === 0) {
        return false;
      }

      const checkbox = candidateRow.locator("span.checkbox-input").first();
      await checkbox.scrollIntoViewIfNeeded();
      await checkbox.click();
      const className = await checkbox.getAttribute('class');
      LoggerUtil.info(`First employee having role ${role} and status ${status} is selected`);
      return className?.includes('focus') ?? false;
    } catch (error) {
      LoggerUtil.error(`Could not select employee having role ${role} whose status is ${status}`, error);
      return false;
    }
  }

  public async clickOnDeleteSelectedUser(): Promise<boolean> {
    try {
      await this.page.locator(AdminPage.DELETE_SELECTED_BTN).first().click({ timeout: 10000 });
      LoggerUtil.info('Clicked on Delete Selected button');
      return true;
    } catch (error) {
      LoggerUtil.error('Could not click on Delete Selected button', error);
      return false;
    }
  }

  public async clickCancelBtnOnDeleteUserModal(): Promise<boolean> {
    try {
      await this.page.locator(AdminPage.MODAL_CANCEL_BTN).first().click({ timeout: 10000 });
      LoggerUtil.info('Clicked on Cancel button on Modal popup');
      return true;
    } catch (error) {
      LoggerUtil.error('Could not click on Cancel button on Modal popup', error);
      return false;
    }
  }
}

File: selenium-framework/src/test/java/com/selenium/testng/web/pages/AdminPage.java
package com.selenium.testng.web.pages;

import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;

import com.selenium.testng.utils.LocatorUtil;
import com.selenium.testng.utils.LoggerUtil;
import com.selenium.testng.web.base.BasePage;
import com.selenium.testng.web.locators.AdminPageLocators;

public class AdminPage extends BasePage {
	
	private static final Logger log = LoggerUtil.getLogger(AdminPage.class);

    public AdminPage(WebDriver driver) {
        super(driver);
    }
  
    By DeleteSelectedBtn = By.xpath(AdminPageLocators.DELETE_SELECTED_BTN);
    By ModalCancelBn = By.xpath(AdminPageLocators.MODAL_CANCEL_BTN);
    
	// ==========================================
	// FUNCTIONS
	// ==========================================
    
	/**
	 * Selects the mentioned first user having the employee role and a status match
	 * We get elements to fetch all user matching with mentioned details 
	 * @param role [String] : Employee's role
	 * @param status [String] : Their status
	 * @return true if user is selected
	 * @author Arzoo Hingorani
	 */
	public boolean selectFirstUser(String role, String status) {
		By usernameChk = LocatorUtil.xpath(AdminPageLocators.USERNAME_CHK, role, status);
		try {
			List<WebElement> usernamesChk = actions.getElements(usernameChk);
			if (usernamesChk != null) {
				actions.scrollIntoView(usernamesChk.get(0));
				actions.click(usernamesChk.get(0));
				log.info("First employee having role {} and status {} is selected", role, status);
				
				if(actions.getAttribute(usernamesChk.get(0), "class").contains("focus"))
					return true;
			}
		} catch (Exception e) {
			log.error("Could not select employee having role {} whose status is {}", role, status, e);
		}
		return false;
	}
	
	/**
	 * Click on Delete Selected user button
	 * @return true if button is clicked
	 * @author Arzoo Hingorani
	 */
	public boolean clickOnDeleteSelectedUser() {
		try {
			if (actions.waitForClickable(DeleteSelectedBtn) != null) {
				actions.scrollIntoView(DeleteSelectedBtn);
				actions.click(DeleteSelectedBtn);
				log.info("Clicked on Delete Selected button");
				return true;
			}
		} catch (Exception e) {
			log.error("Could not click on Delete Selected button", e);
		}
		return false;
	}
	
	/**
	 * Click on Cancel button on the Delete selected user popup
	 * @return true if button is clicked
	 * @author Arzoo Hingorani
	 */
	public boolean clickCancelBtnOnDeleteUserModal() {
		try {
			if (actions.waitForClickable(ModalCancelBn) != null) {
				actions.click(ModalCancelBn);
				log.info("Clicked on Cancel button on Modal popup");
				return true;
			}
		} catch (Exception e) {
			log.error("Could not click on Cancel button on Modal popup", e);
		}
		return false;
	}
}